EIPS/eip-7907.md

```diff
@@ -0,0 +1,59 @@
+---
+eip: 7907
+title: Remove Contract Code Size Limit
+description: Removes the contract code size limit introduced in EIP-170 and adds a gas metering to code loading
+author: Charles Cooper (@charles-cooper)
+discussions-to: https://ethereum-magicians.org/t/eip-remove-contract-size-limit/23156
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-03-14
+requires: 170
+---
+
+## Abstract
+
+This EIP removes the hard contract code size limit of 24KB (24576 bytes) introduced in [EIP-170](./eip-170.md) and replaces it with a gas-based mechanism. It introduces a gas cost of 2 gas per (32 byte) word for contract code exceeding 24KB, allowing deployment of contracts of any size while preventing DoS attacks through appropriate gas metering.
+
+## Motivation
+
+EIP-170 introduced a 24KB contract code size limit to prevent potential DoS attacks, as large contract code requires O(n) resource cost in terms of disk reads, VM preprocessing, and Merkle proof sizes, all of which are not directly compensated by gas fees. However, this hard limit restricts legitimate use cases for large contracts.
+
+Rather than imposing an arbitrary size limit, this EIP proposes a gas-based solution that allows contracts of any size while ensuring that users loading very large contracts pay gas proportional to the additional resources they consume. This approach aligns with Ethereum's gas model philosophy of paying for the resources consumed.
```

Comment:
One side effect of unlimited size of contract code is that it may break the p2p sync assumption - the large size contract code may not fit in a single p2p packet. We found such an issue in geth ~2 years ago.  It may be worth double-checking the issue in the latest clients.

Reply:
the p2p packet size is an order of magnitude more than 24kb. Even just double or tripling this, without breaking anything else, would be a huge improvement for the developer experience. 

We could spend a year researching the downstream effects of making it infinite, and optimizing gas costs, etc..

If there's a path to even just 1.5X'ing the limit without making much bigger protocol changes we need to do that ASAP.

Reply:
> One side effect of the unlimited size of contract code is that it may break the p2p sync assumption - a large-size contract code may not fit in a single p2p packet. We found such an issue in geth ~2 years ago. It may be worth double-checking the issue in the latest clients.

How big was the packet? Because of the 200 gas per byte cost to deploy a contract, there is a code size limit implied by the transaction gas limit (currently around ~170kb).

Reply:
devp2p eth wire packets can be up to [10 MB](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#basic-operation) so I wouldn't be too worried about that.

However it does seem reasonable to set an explicit upper bound even if it isn't strictly necessary. The reason being, if we want to increase the gas limit, we want to minimize the number of things that scale with it.

Reply:
i'm willing to add an upper bound if it's a hard requirement from ACD. however, my hope is that we don't need another EIP + hard fork every time that we want to increase contract sizes, but rather that it scales naturally as gas limit increases.

i do understand though the peace of mind that comes with limits in the protocol. if we do add a limit, i would want to add to the EIP what considerations it is bounded by (disk space assumptions on the client? p2p?) so that when it comes up for discussion, people know exactly what variables were considered during this EIP, rather than letting them guess about "unknown unknowns".

Reply:
[EIP-7825](https://eips.ethereum.org/EIPS/eip-7825) may provide a de-facto limit. But we may want to add a hard coded limit anyway that is a function of that to ensure the code deposit can be paid.  Otherwise it could be griefed by initcode.  

Perhaps 63/64 of the tx gas cap divided by the code deposit cost per byte?  that would be 147,656 bytes or just under 144.2K.  A round binary number would be 147,456 (0x24000 or 144K, or 6x the current limit).

Reply:
> devp2p eth wire packets can be up to [10 MB](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#basic-operation) so I wouldn't be too worried about that.
> 
> However it does seem reasonable to set an explicit upper bound even if it isn't strictly necessary. The reason being, if we want to increase the gas limit, we want to minimize the number of things that scale with it.

It looks like SnapSync is using a much lower number ([512KB](https://github.com/ethereum/go-ethereum/blob/7fed9584b5426be5db6d7b0198acdec6515d9c81/eth/protocols/snap/sync.go#L56)).  Although it may not affect current L1 (max ~170KB), it may kill an L2, whose block limit can easily be higher than 100M (or ~ 500KB code size), like Base is using.

Reply:
does 256KB seem reasonable?

Reply:
256KB sounds to be a good number that should satisfy most contract development in the future.

Reply:
1177bada71fc57eb5bb272ae45267eb0dfb85d39